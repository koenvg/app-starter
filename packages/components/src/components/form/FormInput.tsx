import React, { SFC } from "react";
import { TextFieldProps, TextField } from "@material-ui/core";
import { FieldProps, getIn } from "formik";

export type Props = FieldProps & Omit<TextFieldProps, "error" | "name" | "onChange" | "value">;

const fieldToTextField = ({ field, form, disabled, ...props }: Props): TextFieldProps => {
  const { name } = field;
  const { touched, errors, isSubmitting } = form;

  const fieldError = getIn(errors, name);
  const showError = getIn(touched, name) && !!fieldError;

  return {
    variant: "outlined",
    fullWidth: true,
    label: props.placeholder,
    margin: "normal",
    ...field,
    ...props,
    error: showError,
    helperText: showError ? fieldError : props.helperText,
    disabled: disabled != undefined ? disabled : isSubmitting,
  };
};

export const FormInput: SFC<Props> = ({ children, ...props }) => {
  return <TextField {...fieldToTextField(props)}>{children}</TextField>;
};

FormInput.displayName = "FormInput";
