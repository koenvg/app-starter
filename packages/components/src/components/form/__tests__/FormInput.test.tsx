import { FormInput } from "../FormInput";
import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { Formik, Field, Form } from "formik";
import * as Yup from "yup";

describe("FormInput", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const schema = Yup.object().shape({ name: Yup.string().required("Required") });
  const SimpleForm = ({ disabled = undefined }) => {
    return (
      <Formik onSubmit={jest.fn()} initialValues={{ name: "" }} validationSchema={schema}>
        <Form>
          <Field name="name" placeholder="Name" component={FormInput} disabled={disabled} />
        </Form>
      </Formik>
    );
  };

  it("should render the input container which is rendered by formik", () => {
    const { container } = render(<SimpleForm />);
    expect(container).toMatchSnapshot();
  });

  it("should disable the input field when the prop is provided", () => {
    const { container } = render(<SimpleForm disabled />);
    const name = container.querySelector("input");
    expect(name.disabled).toBeTruthy();
  });

  it("should disable the input field while submitting", async () => {
    const { container } = render(<SimpleForm />);

    const name = container.querySelector("input");
    await wait(() => {
      fireEvent.change(name, {
        target: {
          value: "SomeName",
        },
      });
    });
    await wait(() => fireEvent.submit(container.querySelector("form")));
    expect(name.disabled).toBeTruthy();
  });

  it("should show errors when the field contains errors", async () => {
    const { container, queryByText } = render(<SimpleForm />);

    await wait(() => fireEvent.submit(container.querySelector("form")));
    expect(queryByText("Required")).not.toBeNull();
    expect(container).toMatchSnapshot();
  });
});
