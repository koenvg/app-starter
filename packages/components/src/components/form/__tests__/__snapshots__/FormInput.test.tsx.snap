// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput should render the input container which is rendered by formik 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
        data-shrink="false"
      >
        Name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-61 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-63"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
</div>
`;

exports[`FormInput should show errors when the field contains errors 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-error Mui-error"
        data-shrink="false"
      >
        Name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-invalid="true"
          class="MuiInputBase-input MuiOutlinedInput-input"
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-253 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-255"
          >
            <span>
              Name
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error"
      >
        Required
      </p>
    </div>
  </form>
</div>
`;
